src/interop/js-interop/js-types.md

```diff
@@ -0,0 +1,189 @@
+---
+title: JS Types
+description: Usage information about the core types in JS interop.
+---
+
+## Motivation
+
+Dart values and JS values belong to separate domains. When compiling to Wasm,
+they belong in separate *runtimes* as well. As such, users should treat JS
+values as foreign types. In order to provide types for JS values, we use the set
+of JS types available in `dart:js_interop`. These are extension types prefixed
+with `JS` and are called "JS types".
+
+Importantly, these types have compiler-specific representation types. This means
+that their runtime type will differ based on whether code is compiled to Wasm or
+JS. In order to interact with and examine these JS values, we use `external`
+interop members.
+
+## Type Hierarchy
+
+JS types form a natural type hierarchy:
+
+- Top type: `JSAny`, which is any non-nullish JS value
+  - Primitives: `JSNumber`, `JSBoolean`, `JSString`
+  - `JSSymbol`
+  - `JSBigInt`
+  - `JSObject`, which is any JS object
+    - `JSFunction`
+      - `JSExportedDartFunction`, which represents a Dart callback that was
+      converted to a JS function
+    - `JSArray`
+    - `JSPromise`
+    - `JSDataView`
+    - `JSTypedArray`
+      - JS typed arrays e.g. `JSUint8Array`
+    - `JSBoxedDartObject`, which allows users to box and pass Dart values
+      opaquely within the same Dart runtime
+
+Note that these JS types are *also* interop types. As such, users can and should
+use them as representation types of other interface types. Users will likely
+want to use `JSObject` for this purpose.
```

Comment:
> I'm not sure what this is trying to say. So far, it seems like the page defines JS types as a mechanism made up in dart:js_interop to...interop with JS types. So what's an "interop type" if not just another way to describe a JS type? Is it like, you use these same types/their same exact names in your interop code?

I was hoping to define "interop type" in the syntax section as it's a better fit, but I can see that this reads as super confusing without the additional context. My goal is to communicate some distinction between the types declared in `dart:js_interop` (which we refer to as JS types) and all interop types that you can use (which are transitively extension types on JS types/`@staticInterop` types/maybe `dart:html` types as well as `@staticInterop` types).

As an example:

```dart
import 'dart:js_interop';
import 'dart:html' show Window;

extension type InteropExtensionType(JSObject o) {}

extension type InteropExtensionType2(InteropExtensionType o) {}

extension type InteropExtensionType3(StaticInteropType o) {}

extension type InteropExtensionType4(Window o) {}

@JS()
@staticInterop
class StaticInteropType {}
```

JS types: `JSString`, `JSNumber`, all the other types in `dart:js_interop`
Interop types: `JSString`, `JSNumber`, all the other types in `dart:js_interop`, `InteropExtensionType`, `InteropExtensionType2`, `InteropExtensionType3`, `InteropExtensionType4`, `StaticInteropType`, `Window`

I *think* what I should do instead is remove these comments and have an "interop type" section in the syntax which refers to "JS types". In that same section, I'll likely want to mention why using `JSObject` makes sense.

Reply:
Thanks for explaining! That makes a lot of sense and yeah sorry since this is all coming out non-linearly I should've expected that everything won't be spelled out here. 

Removing the text works fine. But also, with this context, I do actually like your original text now, since it would point to an explanation. But this part: 
> "...use them as representation types of other interface types. Users will likely
want to use `JSObject` for this purpose." 

...Would have to be repeated and elaborated on over there. Up to you; you can decide which is better.

Wherever you define interop types, though, I hope you use something like your example in the comment (plus the list separating which is which afterwards); that really helped me make sense of it!

Reply:
> ...Would have to be repeated and elaborated on over there. Up to you; you can decide which is better.

Good point. I think I'll wait and see when I write the section on syntax - it may end up that here is the better fit for that context.

Noted on using an example/better differentiating JS types and interop types!